File: packages/server/cmd/renewlet/subscription_renewal_test.go
package main

// 续订算法 Go 测试读取 shared JSON fixture；自动/手动续订的 date-only 推进语义必须追随 TypeScript 事实源。

import (
	"encoding/json"
	"os"
	"testing"
)

type subscriptionRenewalFixture struct {
	Name                    string                   `json:"name"`
	Input                   subscriptionRenewalInput `json:"input"`
	Today                   string                   `json:"today"`
	Mode                    renewalMode              `json:"mode"`
	Eligible                bool                     `json:"eligible"`
	ExpectedNextBillingDate string                   `json:"expectedNextBillingDate"`
	ExpectedStatus          string                   `json:"expectedStatus"`
}

func TestSubscriptionRenewalMatchesSharedFixtures(t *testing.T) {
	// Go 版算法不单独发明期望；它必须追 shared fixture，才能证明 Docker 与 Cloudflare 续订口径一致。
	data, err := os.ReadFile("../../../shared/src/subscription-renewal-fixtures.json")
	if err != nil {
		t.Fatalf("read shared fixtures: %v", err)
	}
	var fixtures []subscriptionRenewalFixture
	if err := json.Unmarshal(data, &fixtures); err != nil {
		t.Fatalf("decode shared fixtures: %v", err)
	}
	for _, fixture := range fixtures {
		t.Run(fixture.Name, func(t *testing.T) {
			eligible := isAutoRenewEligible(fixture.Input, fixture.Today)
			if fixture.Mode == renewalModeManual {
				eligible = isManualRenewEligible(fixture.Input)
			}
			if eligible != fixture.Eligible {
				t.Fatalf("eligible=%v, want %v", eligible, fixture.Eligible)
			}

			result, ok, err := advanceSubscriptionRenewal(fixture.Input, fixture.Today, fixture.Mode)
			if err != nil {
				t.Fatalf("advance: %v", err)
			}
			if ok != fixture.Eligible {
				t.Fatalf("advanced=%v, want %v", ok, fixture.Eligible)
			}
			if !fixture.Eligible {
				return
			}
			if result.NextBillingDate != fixture.ExpectedNextBillingDate || result.Status != fixture.ExpectedStatus {
				t.Fatalf("result=%+v, want next=%s status=%s", result, fixture.ExpectedNextBillingDate, fixture.ExpectedStatus)
			}
		})
	}
}

File: packages/shared/src/subscription-renewal.test.ts
// 续订算法测试读取共享 JSON fixture；同一批样例也由 Go 后端读取，用来锁住跨运行面账单日推进语义。
import { describe, expect, it } from "vitest";
import fixtures from "./subscription-renewal-fixtures.json";
import {
  advanceSubscriptionRenewal,
  isAutoRenewEligible,
  isManualRenewEligible,
  type RenewalMode,
  type SubscriptionRenewalInput,
} from "./subscription-renewal";

type Fixture = {
  name: string;
  input: SubscriptionRenewalInput;
  today: string;
  mode: RenewalMode;
  eligible: boolean;
  expectedNextBillingDate?: string;
  expectedStatus?: string;
};

describe("subscription renewal", () => {
  // 这份 fixture 同时被 Go 后端读取；新增续订规则时先扩展 fixture，再让两端实现追同一组期望。
  it.each(fixtures as Fixture[])("matches fixture $name", (fixture) => {
    const eligible = fixture.mode === "auto"
      ? isAutoRenewEligible(fixture.input, fixture.today)
      : isManualRenewEligible(fixture.input);
    expect(eligible).toBe(fixture.eligible);

    const result = advanceSubscriptionRenewal(fixture.input, fixture.today, fixture.mode);
    if (!fixture.eligible) {
      expect(result).toBeNull();
      return;
    }
    expect(result).toEqual({
      nextBillingDate: fixture.expectedNextBillingDate,
      status: fixture.expectedStatus,
    });
  });
});
